cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala

```diff
@@ -0,0 +1,153 @@
+
+  def getAddress: Seq[String] = getScalarParam(address)
+
+  def setAddress(v: Seq[String]): this.type = setScalarParam(address, v)
+
+  def setAddress(v: String): this.type = setScalarParam(address, Seq(v))
+
+  def getAddressCol: String = getVectorParam(address)
+
+  def setAddressCol(v: String): this.type = setVectorParam(address, v)
+
+  protected def prepareEntity: Row => Option[AbstractHttpEntity]
+
+  protected def prepareMethod(): HttpRequestBase = new HttpGet()
+
+  protected def contentType: Row => String = { _ => "application/json" }
+
+  protected def inputFunc(schema: StructType): Row => Option[HttpRequestBase] = {
+    { row: Row =>
+      if (shouldSkip(row)) {
+        None
+      } else {
+        val req = prepareMethod()
+        val addressValue = getValueOpt(row, address).mkString("")
+        val queryParams = "?" + URLEncodingUtils.format(Map("api-version" -> "1.0",
+          "subscription-key" -> getSubscriptionKey,
+          "query" -> addressValue))
+        req.setURI(new URI(getUrl + queryParams))
+
+        Some(req)
+      }
+    }
+  }
+
+  protected def getInternalInputParser(schema: StructType): HTTPInputParser = {
+    new CustomInputParser().setNullableUDF(inputFunc(schema))
+  }
+}
+
+trait HasBatchAddressInput extends HasServiceParams with HasSubscriptionKey with HasURL {
+  val address = new ServiceParam[Seq[String]](
+    this, "address", "the address to geocode")
+
+  def getAddresses: Seq[String] = getScalarParam(address)
+
+  def setAddresses(v: Seq[String]): this.type = setScalarParam(address, v)
+
+  def setAddresses(v: String): this.type = setScalarParam(address, Seq(v))
+
+  def getAddressesCol: String = getVectorParam(address)
+
+  def setAddressesCol(v: String): this.type = setVectorParam(address, v)
+
+  protected def prepareEntity: Row => Option[AbstractHttpEntity]
+
+  protected def contentType: Row => String = { _ => "application/json" }
+
+  protected def inputFunc(schema: StructType): Row => Option[HttpRequestBase] = {
+    { row: Row =>
+      if (shouldSkip(row)) {
+        None
+      } else {
+
+        val queryParams = "?" + URLEncodingUtils.format(Map("api-version" -> "1.0",
+          "subscription-key" -> getSubscriptionKey))
+        val post = new HttpPost(new URI(getUrl + queryParams))
+        post.setHeader("Content-Type", "application/json")
+        val addressesCol = getValueOpt(row, address)
+        val encodedAddresses = addressesCol.get.map(x => URLEncoder.encode(x, "UTF-8")).toList
+        val payloadItems = encodedAddresses.map(x => s"""{ "query": "?query=$x&limit=1" }""").mkString(",")
+        val payload = s"""{ "batchItems": [ $payloadItems ] }"""
+        post.setEntity(new StringEntity(payload))
+        Some(post)
+      }
+    }
+  }
+
+  protected def getInternalInputParser(schema: StructType): HTTPInputParser = {
+    new CustomInputParser().setNullableUDF(inputFunc(schema))
+  }
+}
+
+trait BatchAsyncReply extends HasAsyncReply with HasSubscriptionKey {
```

Comment:
Im not seeing much azure maps specific stuff here, is it possible to re-use existing async reply trait to keep it DRY?

Reply:
The primary difference is how auth information is passed. Most Cognitive Services pass in the subscription key as a request header. Also I want to implement response massaging in this trait.